### PR TITLE
Naive light client synchronization

### PR DESCRIPTION
**Pull request by rphmeier**
This work was getting pretty massive so I've decided to cut it off here for now and leave further improvements/tests to future PRs.

Please review but don't merge until 1.5 is released.

### CODE REVIEW

**Comment by rphmeier** [comment]
The CLI is pretty ugly and I'm not a huge fan of having the Informant/Client split -- will do some work to unify as much as possible via traits.

**Comment by arkpar** [comment]
ethcore/res/ethereum/tests needs reverting

File: ethcore/light/src/client/header_chain.rs [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+		// hold candidates the whole time to guard import order.
+		let mut candidates = self.candidates.write();
+
+		// find parent details.
+		let parent_td = {
+			if number == 1 {
+				let g_view = HeaderView::new(&self.genesis_header);
+				g_view.difficulty()
+			} else {
```
Comment (NikVolf) [inline ethcore/light/src/client/header_chain.rs]:
```rust
else {
				try!(candidates.get(&(number - 1))
					.and_then(|entry| entry.candidates.iter().find(|c| c.hash == parent_hash))
					.map(|c| c.total_difficulty)
					.ok_or(BlockError::UnknownParent(parent_hash)))
			}
```

?

File: ethcore/light/src/client/header_chain.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+	/// so including it within a CHT would be redundant.
+	pub fn cht_root(&self, n: usize) -> Option<H256> {
+		self.cht_roots.lock().get(n).map(|h| h.clone())
+	}
+
+	/// Get the genesis hash.
+	pub fn genesis_hash(&self) -> H256 {
+		use util::Hashable;
+
+		self.genesis_header.sha3()
```
Comment (NikVolf) [inline ethcore/light/src/client/header_chain.rs]: one-liner
`::util::Hashable::sha3(&self.genesis_header)`
?

**Comment by rphmeier** [comment]
Given that 1.5 has been delayed a bit, it might make sense to remove the CLI option for now and merge it into master afterwards. I have a few pending PRs which build upon it but I haven't made yet in anticipation of the merge.

**Comment by rphmeier** [comment]
(CLI removed)

**Pull request merged**